### ISSUE TITLE
PowerMac and error: 'kCCSuccess' undeclared (first use in this function)

### ISSUE THREAD

**Issue by noloader** [openssl/openssl#16408]
I'm building the latest OpenSSL and OpenSSH on a PowerMac G5. I'm using OpenSSL 1.1.1l.

Make is failing with:

```
cc  -I. -Iinclude -fPIC -fno-common -arch ppc -std=gnu9x -Wa,-force_cpusubtype_ALL -g2 -O2 -mlong-double-64 -fno-common -maltivec -fPIC -pthread -DB_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_BN_ASM_MONT -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DAES_ASM -DVPAES_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/opt/ssh\"" -DENGINESDIR="\"/opt/ssh/lib/engines-1.1\"" -D_REENTRANT -DNDEBUG -I/opt/ssh/include -DNDEBUG -DPEDANTIC  -c -o crypto/rand/rand_unix.o crypto/rand/rand_unix.c
crypto/rand/rand_unix.c:38:40: error: CommonCrypto/CommonRandom.h: No such file or directory
crypto/rand/rand_unix.c: In function 'syscall_random':
crypto/rand/rand_unix.c:385: warning: implicit declaration of function 'CCRandomGenerateBytes'
crypto/rand/rand_unix.c:385: error: 'kCCSuccess' undeclared (first use in this function)
crypto/rand/rand_unix.c:385: error: (Each undeclared identifier is reported only once
crypto/rand/rand_unix.c:385: error: for each function it appears in.)
```

Also see [GH #16248](https://github.com/openssl/openssl/issues/16248).

Here is the configdata file produced during config: [configdata.pm.zip](https://github.com/openssl/openssl/files/7041997/configdata.pm.zip).

**Comment by ericonr** [comment]
Basically duplicate of #16407

**Comment by noloader** [comment]
I added the include `<CommonCrypto/CommonCryptoError.h>` per GH #16248, but no joy. That generates two warnings for missing headers and still leads to a failed compile.

```
cc  -I. -Iinclude -fPIC -fno-common -arch ppc -std=gnu9x -Wa,-force_cpusubtype_ALL -g2 -O2 -mlong-double-64 -fno-common -maltivec -fPIC -pthread -DB_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_BN_ASM_MONT -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DAES_ASM -DVPAES_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/opt/ssh\"" -DENGINESDIR="\"/opt/ssh/lib/engines-1.1\"" -D_REENTRANT -DNDEBUG -I/opt/ssh/include -DNDEBUG -DPEDANTIC  -c -o crypto/rand/rand_unix.o crypto/rand/rand_unix.c
crypto/rand/rand_unix.c:38:40: error: CommonCrypto/CommonRandom.h: No such file or directory
crypto/rand/rand_unix.c:39:45: error: CommonCrypto/CommonCryptoError.h: No such file or directory
crypto/rand/rand_unix.c: In function 'syscall_random':
crypto/rand/rand_unix.c:386: warning: implicit declaration of function 'CCRandomGenerateBytes'
crypto/rand/rand_unix.c:386: error: 'kCCSuccess' undeclared (first use in this function)
crypto/rand/rand_unix.c:386: error: (Each undeclared identifier is reported only once
crypto/rand/rand_unix.c:386: error: for each function it appears in.)
make[1]: *** [crypto/rand/rand_unix.o] Error 1
```

**Comment by debohman** [comment]
> I added the include `<CommonCrypto/CommonCryptoError.h>` per GH #16248, but no joy. That generates two warnings for missing headers and still leads to a failed compile.
> 
> ```
> cc  -I. -Iinclude -fPIC -fno-common -arch ppc -std=gnu9x -Wa,-force_cpusubtype_ALL -g2 -O2 -mlong-double-64 -fno-common -maltivec -fPIC -pthread -DB_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_BN_ASM_MONT -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DAES_ASM -DVPAES_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/opt/ssh\"" -DENGINESDIR="\"/opt/ssh/lib/engines-1.1\"" -D_REENTRANT -DNDEBUG -I/opt/ssh/include -DNDEBUG -DPEDANTIC  -c -o crypto/rand/rand_unix.o crypto/rand/rand_unix.c
> crypto/rand/rand_unix.c:38:40: error: CommonCrypto/CommonRandom.h: No such file or directory
> crypto/rand/rand_unix.c:39:45: error: CommonCrypto/CommonCryptoError.h: No such file or directory
> crypto/rand/rand_unix.c: In function 'syscall_random':
> crypto/rand/rand_unix.c:386: warning: implicit declaration of function 'CCRandomGenerateBytes'
> crypto/rand/rand_unix.c:386: error: 'kCCSuccess' undeclared (first use in this function)
> crypto/rand/rand_unix.c:386: error: (Each undeclared identifier is reported only once
> crypto/rand/rand_unix.c:386: error: for each function it appears in.)
> make[1]: *** [crypto/rand/rand_unix.o] Error 1
> ```

Which release of macOS? Looks like the CommonCrypto headers do not exist.

**Comment by noloader** [comment]
@debohman,

> Which release of macOS? Looks like the CommonCrypto headers do not exist.

OS X 10.5 on an old PowerMac (circa 2002). That old PowerMac still beats the snot out of Intel Atoms and most Celerons. I really enjoy working on it.

This is the patch I used.

```diff
--- crypto/rand/rand_unix.c
+++ crypto/rand/rand_unix.c
@@ -34,8 +34,9 @@
 #if defined(__OpenBSD__)
 # include <sys/param.h>
 #endif
-#if defined(__APPLE__)
+#if defined(__APPLE__) && !defined(OPENSSL_NO_CCRANDOMGENERATEBYTES)
 # include <CommonCrypto/CommonRandom.h>
+# include <CommonCrypto/CommonCryptoError.h>
 #endif
 
 #if defined(OPENSSL_SYS_UNIX) || defined(__DJGPP__)
@@ -381,7 +382,7 @@
         if (errno != ENOSYS)
             return -1;
     }
-#  elif defined(__APPLE__)
+#  elif defined(__APPLE__) && !defined(OPENSSL_NO_CCRANDOMGENERATEBYTES)
     if (CCRandomGenerateBytes(buf, buflen) == kCCSuccess)
 	    return (ssize_t)buflen;
```

I then add `-DOPENSSL_NO_CCRANDOMGENERATEBYTES` to `CPPFLAGS` when building on OS X 10.5 and below.

**Comment by debohman** [comment]
Yes, I mentioned in #16407 that CommonCrypto/CommonRandom.h seems to be a later extension to the API. I checked an old Mavericks root, and the header file is not present.

**Comment by t8m** [comment]
Closing as duplicate of #16407
